--- a/libavformat/hlsenc.c
+++ b/libavformat/hlsenc.c
@@ -246,6 +246,7 @@
     ClosedCaptionsStream *cc_streams;
     unsigned int nb_ccstreams;
 
+    int pts_alignment_enabled; /* use start_pts to calculate segment duration */
     int master_m3u8_created; /* status of master play-list creation */
     char *master_m3u8_url; /* URL of the master m3u8 file */
     int version; /* HLS version */
@@ -2448,12 +2449,37 @@
 
     return ret;
 }
+
+// Determines whether to enable alternative end_pts calculation.
+// Input seek position and start_number should match a m3u8 file already generated.
+static int end_pts_resume_from_start_number(VariantStream *vs, AVStream *st, HLSContext *hls)
+{
+    if (vs->start_pts <= 0 ||
+        vs->start_pts == AV_NOPTS_VALUE ||
+        hls->start_sequence == 0 ||
+        (hls->flags & HLS_SPLIT_BY_TIME) ||
+        hls->init_time > 0) {
+        return AVERROR(EINVAL);
+    }
+
+    if (hls->start_sequence > INT64_MAX / hls->recording_time) {
+        return AVERROR(EOVERFLOW);
+    }
+    int64_t start_sequence_time = hls->start_sequence * hls->recording_time;
+    // int64_t start_pts_us = av_rescale_q(vs->start_pts, st->time_base, AV_TIME_BASE_Q);
+    if (av_compare_ts(vs->start_pts, st->time_base, start_sequence_time, AV_TIME_BASE_Q) < 0)
+        return AVERROR(EINVAL);
+    // if (start_pts_us < start_sequence_time)
+
+    return 0;
+}
+
 static int hls_write_packet(AVFormatContext *s, AVPacket *pkt)
 {
     HLSContext *hls = s->priv_data;
     AVFormatContext *oc = NULL;
     AVStream *st = s->streams[pkt->stream_index];
-    int64_t end_pts = 0;
+    int64_t end_pts = 0, pts = 0;
     int is_ref_pkt = 1;
     int ret = 0, can_split = 1, i, j;
     int stream_index = 0;
@@ -2505,7 +2531,16 @@
         vs->start_pts = pkt->pts;
         if (st->codecpar->codec_type == AVMEDIA_TYPE_AUDIO)
             vs->start_pts_from_audio = 1;
+
+        // If enabled, calculate segment lengths based on start_sequence and start_pts.
+        // Improves segment alignment when seeking for static manifests during stream copy.
+        if (!end_pts_discont_from_start(vs, st, hls)) {
+            hls->pts_alignment_enabled = 1;
+            av_log(s, AV_LOG_INFO, "Using segment length calculation based on start time and HLS_START_SEQUENCE=%"PRId64".\n",
+                   hls->start_sequence);
+        }
     }
+
     if (vs->start_pts_from_audio && st->codecpar->codec_type == AVMEDIA_TYPE_VIDEO && vs->start_pts > pkt->pts) {
         vs->start_pts = pkt->pts;
         vs->start_pts_from_audio = 0;
@@ -2538,8 +2573,16 @@
         }
     }
 
+    pts = pkt->pts - vs->start_pts;
+
+    if (hls->pts_alignment_enabled) {
+        pts = pkt->pts;
+        /* Calculate end_pts as if segments ahead of start_sequence has already been written  */
+        end_pts = hls->recording_time * (vs->number + hls->start_sequence);
+    }
+
     can_split = can_split && (pkt->pts - vs->end_pts > 0);
-    if (vs->packets_written && can_split && av_compare_ts(pkt->pts - vs->start_pts, st->time_base,
+    if (vs->packets_written && can_split && av_compare_ts(pts, st->time_base,
                                                           end_pts, AV_TIME_BASE_Q) >= 0) {
         int64_t new_start_pos;
         int byterange_mode = (hls->flags & HLS_SINGLE_FILE) || (hls->max_seg_size > 0);
--- a/libavformat/movenc.c
+++ b/libavformat/movenc.c
@@ -8547,6 +8547,7 @@
     { AV_CODEC_ID_HEVC,            MKTAG('h', 'e', 'v', '1') },
     { AV_CODEC_ID_HEVC,            MKTAG('h', 'v', 'c', '1') },
     { AV_CODEC_ID_HEVC,            MKTAG('d', 'v', 'h', '1') },
+    { AV_CODEC_ID_HEVC,            MKTAG('d', 'v', 'h', 'e') },
     { AV_CODEC_ID_VVC,             MKTAG('v', 'v', 'c', '1') },
     { AV_CODEC_ID_VVC,             MKTAG('v', 'v', 'i', '1') },
     { AV_CODEC_ID_EVC,             MKTAG('e', 'v', 'c', '1') },
@@ -8560,6 +8561,7 @@
     { AV_CODEC_ID_TSCC2,           MKTAG('m', 'p', '4', 'v') },
     { AV_CODEC_ID_VP9,             MKTAG('v', 'p', '0', '9') },
     { AV_CODEC_ID_AV1,             MKTAG('a', 'v', '0', '1') },
+    { AV_CODEC_ID_AV1,             MKTAG('d', 'a', 'v', '1') },
     { AV_CODEC_ID_AAC,             MKTAG('m', 'p', '4', 'a') },
     { AV_CODEC_ID_ALAC,            MKTAG('a', 'l', 'a', 'c') },
     { AV_CODEC_ID_MP4ALS,          MKTAG('m', 'p', '4', 'a') },
